### PR TITLE
Make destructors on `extern "C"` frames to be executed

### PR DESCRIPTION
**Pull request by nbdd0121**
This would make the example in #123231 print "Noisy Drop". I didn't mark this as fixing the issue because the behaviour is yet to be spec'ed.

Tracking:

- https://github.com/rust-lang/rust/issues/74990

### CODE REVIEW

**Comment by tmandry** [comment]
Any ideas for how we could measure the cost of this change in real-world code?

**Comment by arielb1** [comment]
What is actually supposed to be causing a cost in this change, other than generating slightly more landing pads?

**Comment by nbdd0121** [comment]
Mostly size cost (landing pads). I guess it doesn't hurt to do a perf run?

**Comment by arielb1** [comment]
Are there that many extern C function to cause a serious size cost?

**Comment by tmandry** [comment]
Am I correct in thinking that there will be no cost if built with panic=abort, it will still abort immediately?

**Comment by arielb1** [comment]
> Am I correct in thinking that there will be no cost if built with panic=abort, it will still abort immediately?

The landing pad change does not affect panic=abort

File: compiler/rustc_mir_transform/src/abort_unwinding_calls.rs [hunk 1]
```diff
@@ -88,17 +98,18 @@ impl<'tcx> MirPass<'tcx> for AbortUnwindingCalls {
...
                 continue;
             }
 
             // Otherwise if this function can unwind, then if the outer function
             // can also unwind there's nothing to do. If the outer function
-            // can't unwind, however, we need to change the landing pad for this
-            // function call to one that aborts.
-            if !body_can_unwind {
+            // can't unwind, however, we need to ensure that any `UnwindAction::Continue`
+            // is replaced with terminate. For those with cleanup blocks, cleanup will still happen,
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/abort_unwinding_calls.rs:110]:
```suggestion
            // is replaced with terminate. For those with UnwindAction::Cleanup, cleanup will still happen,
```

**Comment by RalfJung** [comment]
It is concerning that CI just passes with hardly any test changes. Could you add a  ui run-pass test with the example from https://github.com/rust-lang/rust/issues/123231? You can model it after `tests/ui/panics/panic-in-cleanup.rs`.

And maybe we should have a mir-opt test as well so that one can see the `UnwindTerminate` terminators being generated.

**Comment by RalfJung** [comment]
If we want to go ahead with this, what do we do with beta (about to become 1.81 real soon)? I doubt we want to land a change like this so late in the cycle.

@Mark-Simulacrum can we still land a PR that reverts `feature(c_unwind)` stabilization (https://github.com/rust-lang/rust/pull/116088) on the beta branch or is that too disruptive?

**Comment by nbdd0121** [comment]
Can we document the current behaviour (prefix of frames) for 1.81 in the relnotes and then we further specify that it's all-or-nothing in 1.82?

**Comment by nbdd0121** [comment]
> It is concerning that CI just passes with hardly any test changes. Could you add a ui run-pass test with the example from #123231? You can model it after `tests/ui/panics/panic-in-cleanup.rs`.
> 
> And maybe we should have a mir-opt test as well so that one can see the `UnwindTerminate` terminators being generated.

Thanks for the pointer. I'll add that when I have time.

**Comment by RalfJung** [comment]
In fact you can just edit `tests/ui/panics/panic-in-ffi.rs` to add a type with destructor in `panic_in_ffi`.

**Comment by Mark-Simulacrum** [comment]
I don't think there's any blockers to landing PRs on beta doing anything at this point, modulo T-compiler signoff. Reverting things to status quo on stable feels pretty safe to me.

**Comment by traviscross** [comment]
@rustbot +T-lang +I-lang-nominated

Over here:

- https://github.com/rust-lang/rust/issues/123231#issuecomment-2312659338

@RalfJung proposes a particular semantic that is implemented by this PR:

> I think we should guarantee that we run all destructors during unwind, and leave room for "unwind might fail to initiate and abort immediately instead" to account for 2-phase unwinding. That's reasonably easy to understand. This is currently not the case but #129582 implements that, IIUC.
> 
> I'm not aware of any code that would _rely_ on the abort happening "early", i.e. skipping some destructors. On current stable, the destructor in the OP example actually does run, so the proposed guarantee (implemented by #129582) is also closer to the status quo than what happens in current beta.

Having followed the thread, I agree this makes the most sense, so I propose...

@rfcbot fcp merge

With respect to what to do about Rust 1.81, I probably think it's fine to ship as-is as long as we're clear about the limits of the guarantees that we're making, but I'm also open to a revert if people think that makes more sense.

**Comment by tmandry** [comment]
> @Mark-Simulacrum can we still land a PR that reverts `feature(c_unwind)` stabilization (https://github.com/rust-lang/rust/pull/116088) on the beta branch or is that too disruptive?

I think the only way this would make sense is if it includes the change to the status quo unwinding behavior, does it?

**Comment by traviscross** [comment]
@rfcbot fcp merge

(The bot didn't kick it off the last time.  Let's try again.)

**Comment by nnethercote** [comment]
I don't know anything about this stuff, but given how many people have chimed in with comments, and the fact that the lang-team has approved it, I plan to rubber-stamp this PR once the FCP is over. If anybody has opinions about the code changes, please weigh in (or even take review assignment) before then. Thanks.

**Comment by nnethercote** [comment]
The FCP ended ~3 weeks ago, so I think this is good to go.

@bors r+

**Comment by nbdd0121** [comment]
The failing test `tests/run-make/longjmp-across-rust` looks like UB to me.

**Comment by RalfJung** [comment]
It probably needs `C-unwind`?

longjmp across frames without destructors is IIRC intended to work.

**Comment by nbdd0121** [comment]
Yeah, but there is a destructor here, albeit empty.

**Comment by RalfJung** [comment]
Oh. Yeah that's UB indeed.
Maybe adjust the test to remove the destructor?

Cc @alexcrichton just FYI, a test you added many years ago turns out to have UB... not sure if there's any context there that is still relevant.

**Comment by alexcrichton** [comment]
It's ok times change and we learn things, I think it's a better move to consider unwinding/longjmp over rust frames with destructors as UB rather than "something that should work", so sounds like it's a good test to remove/update 👍

**Comment by BatmanAoD** [comment]
Removing the test is correct; `longjmp` over Rust frames with destructors is always UB, per RFC 2945. Such frames are not [Plane Old Frames](https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md#plain-old-frames), and:

> This RFC specifies that, regardless of the platform or the ABI string (`"C"` or `"C-unwind"`), any platform features that may rely on forced unwinding will always be considered undefined behavior if they cross non-POFs.

`longjmp` is one such platform feature that "may rely on forced unwinding."

**Pull request merged**